### PR TITLE
Improve benchmark suite

### PR DESCRIPTION
**Pull request by dralley**

### CODE REVIEW

**Comment by dralley** [comment]
Benchmarks compile but one of the original ones is still / re- broken

File: benches/macrobenches.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+static RPM_OTHER: &[u8] = include_bytes!("../tests/documents/rpm_other.xml");
+static LIBREOFFICE_DOCUMENT: &[u8] = include_bytes!("../tests/documents/libreoffice_document.fodt");
+static DOCUMENT: &[u8] = include_bytes!("../tests/documents/document.xml");
+static TEST_WRITER_INDENT: &[u8] = include_bytes!("../tests/documents/test_writer_indent.xml");
+static SAMPLE_1: &[u8] = include_bytes!("../tests/documents/sample_1.xml");
+static LINESCORE: &[u8] = include_bytes!("../tests/documents/linescore.xml");
+static SAMPLE_RSS: &[u8] = include_bytes!("../tests/documents/sample_rss.xml");
+static PLAYERS: &[u8] = include_bytes!("../tests/documents/players.xml");
+
+// TODO: read the namespaces too
```
Comment (dralley) [inline benches/macrobenches.rs:18]: No point doing this yet if the API is about to change, which I assume it is.

File: benches/macrobenches.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    });
+
+    // long, mostly text, not much escaping
+    group.bench_function("rpm_filelists.xml", |b| {
+        b.iter(|| {
+            parse_document(RPM_FILELISTS).unwrap();
+        })
+    });
+
+    // long, mix of attributes and text, not much escaping, lots of utf-8 characters, lots of namespaces
```
Comment (dralley) [inline benches/macrobenches.rs:67]: I meant non-ascii characters.  I can fix that

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -49,6 +44,10 @@ jobs:
       env:
         LLVM_PROFILE_FILE: coverage/all-features-%p-%m.profraw
       run: cargo test --all-features
+    - name: Run tests (all features) + benchmarks
```
Comment (dralley) [inline .github/workflows/rust.yml:47]: Only runs the benchmarks with 1 iteration, as though they are tests.  So it checks that they work without taking 5 minutes.

File: benches/microbenches.rs [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+        let src = format!(r#"<hello target="{}">"#, "world".repeat(512 / 5)).into_bytes();
+        let mut buf = Vec::with_capacity(1024);
+        b.iter(|| {
+            let mut r = Reader::from_reader(src.as_ref());
+            let mut nbtxt = criterion::black_box(0);
+            r.check_end_names(false)
+                .check_comments(false)
+                .trim_text(true);
+            match r.read_event(&mut buf) {
+                Ok(Event::Start(ref e)) => nbtxt += e.name().as_ref().len(),
```
Comment (Mingun) [inline benches/microbenches.rs:239]: `BytesStart` implements `Deref<[u8]>`, so why you replaced `e.len()` with `e.name().as_ref().len()`? Not that it was important...

**Review by Mingun — APPROVED** [review]
I spot some minor nits, but it is good enough. Feel free to merge it as is

**Pull request merged**